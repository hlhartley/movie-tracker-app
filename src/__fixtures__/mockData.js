export const mockMovieData = [
    {   
        vote_count: 3119,
        id: 297802,
        video: false,
        vote_average: 6.9,
        title: "Aquaman",
        popularity: 483.065,
        poster_path: "/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg",
        original_language: "en",
        original_title: "Aquaman",
        backdrop_path: "/5A2bMlLfJrAfX9bqAibOL2gCruF.jpg",
        adult: false,
        overview: "Once home to the most advanced civilization on Earth, the city of Atlantis is now an underwater kingdom ruled by the power-hungry King Orm. With a vast army at his disposal, Orm plans to conquer the remaining oceanic people -- and then the surface world. Standing in his way is Aquaman, Orm's half-human, half-Atlantean brother and true heir to the throne. With help from royal counselor Vulko, Aquaman must retrieve the legendary Trident of Atlan and embrace his destiny as protector of the deep.",
        release_date: "2018-12-07",
    },
    {
        vote_count: 1020,
        id: 424783,
        video: false,
        vote_average: 6.5,
        title: "Bumblebee",
        popularity: 263.575,
        poster_path: "/fw02ONlDhrYjTSZV8XO6hhU3ds3.jpg",
        original_language: "en",
        original_title: "Bumblebee",
        backdrop_path: "/hMANgfPHR1tRObNp2oPiOi9mMlz.jpg",
        adult: false,
        overview: "On the run in the year 1987, Bumblebee finds refuge in a junkyard in a small Californian beach town. Charlie, on the cusp of turning 18 and trying to find her place in the world, discovers Bumblebee, battle-scarred and broken. When Charlie revives him, she quickly learns this is no ordinary yellow VW bug.",
        release_date: "2018-12-15",
    },
    { 
        vote_count: 829,
        id: 480530,
        video: false,
        vote_average: 6.6,
        title: "Creed II",
        popularity: 191.233,
        poster_path: "/v3QyboWRoA4O9RbcsqH8tJMe8EB.jpg",
        original_language: "en",
        original_title: "Creed II",
        backdrop_path: "/8yqLPNwNCtpOPc3XkOlkSMnghzw.jpg",
        adult: false,
        overview: "Follows Adonis Creed's life inside and outside of the ring as he deals with new found fame, issues with his family, and his continuing quest to become a champion.",
        release_date: "2018-11-21",
    }
]