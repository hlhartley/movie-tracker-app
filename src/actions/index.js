export const addMovies = (movies) => ({
  type: 'ADD_MOVIES',
  movies
});

export const loginUser = (id, name) => ({
  type: 'LOGIN_USER',
  id,
  name
});

export const updateError = (message) => ({
  type: 'UPDATE_ERROR',
  message
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});

export const showPopup = (bool) => ({
  type: 'SHOW_POPUP',
  showPopup: bool
});

export const toggleFavorite = (id) => ({
  type: 'TOGGLE_FAVORITE',
  id
});
