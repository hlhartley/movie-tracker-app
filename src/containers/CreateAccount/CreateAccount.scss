.form-container {
    display: flex;
    flex-flow: column;
    justify-content: center;
    align-items: center;
    line-height: 55px;
    font-size: 30px;
    font-weight: 700;
    height: 50vh;
    background-color: $brown;
    margin-top: 35px;
    width: 80%;
    border-radius: 5px;
}

.main-text {
    color: $gold;
}

.sub-text {
    font-size: 20px;
    font-weight: 300;
    color: $gold;
}

p {
    margin-top: -15px;
    margin-bottom: 2px;
}

.create-account-btn {
    margin-top: 20px;
    border-radius: 10px;
    width: 22vw;
    border: 1px solid black;
    background-color: $silver;
    font-weight: 700;
    padding: 5px;
    &:hover {
        background-color: $darksilver;
    }
}

input {
    width: 40vw;
    text-align: center;
    margin: 10px;
    margin-left: 20px;
}

.name-container,
.email-container,
.password-container {
    display: flex;
    justify-content: space-around;
}

.fa-user-circle {
    font-size: 60px;
    margin: 25px;
    color: $gold;
}