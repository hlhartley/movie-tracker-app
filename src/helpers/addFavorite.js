export const addFavorite = async (movie_id, user_id, title, poster_path, release_date, vote_average, overview) => {
  const response = await fetch('http://localhost:3000/api/users/favorites/new', {
    method: 'POST',
    body: JSON.stringify({
      movie_id, 
      user_id, 
      title, 
      poster_path, 
      release_date, 
      vote_average, 
      overview
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (response.status >= 300) {
    throw Error(`Error saving to favorites: ${response.statusText}`);
  } else {
    const result = await response.json();
    return result;
  }
}