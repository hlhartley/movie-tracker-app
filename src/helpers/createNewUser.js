export const createNewUser = async (user) => {

  const response = await fetch('http://localhost:3000/api/users/new', {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (response.status >= 300) {
    throw Error(`Email has already been used: ${response.statusText}`);
  } else {
    const result = await response.json();
    return result;
  }

}