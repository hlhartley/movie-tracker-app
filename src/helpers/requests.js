import { apikey } from '../apikey';

export const fetchMovies = async () => {
  const url = `https://api.themoviedb.org/3/movie/now_playing?api_key=${apikey}&language=en-US&page=1`;
  const response= await fetch(url);
  if (response.status >= 300) {
    throw new Error(`Error fetching', code: ${response.status}`)
  } else {
    return response.json();
  }
}