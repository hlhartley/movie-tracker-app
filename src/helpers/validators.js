const nameRegEx = /^[a-zA-Z]+$/ 
const emailRegEx = /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/
const passwordRegEx = /^[a-zA-Z0-9@*#]{2,10}$/ 

export const validator = (type, value) => {
    switch (type) {
        case 'email':
            return emailRegEx.test(value.trim());
        case 'name':
            return nameRegEx.test(value.trim());
        case 'password':
            return passwordRegEx.test(value);
        default:
            return false;
    }
}