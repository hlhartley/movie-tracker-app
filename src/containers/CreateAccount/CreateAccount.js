import React, { Component } from 'react';
import { connect } from 'react-redux';
import { createNewUser } from '../../helpers/requests';
import { Redirect } from 'react-router-dom';
import { updateError, loginUser } from '../../actions';
import { validator } from '../../helpers/validators';

export class CreateAccount extends Component {
    constructor() {
        super();
        this.state = {
            name: '',
            email: '',
            password: '',
        }
    }

    handleChange = (e) => {
        let { name, value} = e.target
        this.setState({ [name]: value })
    }

    handleSubmit = async () => {
        let user = {...this.state};
        user.email = user.email.toLowerCase();
        try {
            const newUser = await createNewUser(user);
            this.props.loginUser(newUser.id, this.state.name)
        } catch(error) {
            this.props.updateError(error.message)
        }
    }

    validateInput = (e) => {
        e.preventDefault();
        this.props.updateError('');
        const { name, email, password } = this.state;
        if (!validator('name', name) || !validator('email', email) || !validator('password', password)) {
            this.props.updateError('Input not successful');
        } else {
            this.handleSubmit();
        }
    }

    render() {
        let { currentUser, errorStatus } = this.props;
        if (currentUser) {
            return(
                <Redirect to='/'/>
            )
        } else {
            return(
                <form onSubmit={this.validateInput} className='form-container'>
                    <p className='main-text'>SIGN UP FOR FREE</p>
                    <p className='sub-text'>Start tracking your favorite movies today!</p>
                    <div className='name-container'>
                        <label htmlFor='name'><i className="fas fa-user"></i></label>
                        <input name='name' value={this.state.name} id='name' onChange={this.handleChange} placeholder='Full name'/>
                    </div>    
                    <div className='email-container'>
                        <label htmlFor='email'><i className="far fa-envelope"></i></label>
                        <input name='email' value={this.state.email} id='email' onChange={this.handleChange} placeholder='E-mail address'/>
                    </div>
                    <div className='password-container'>   
                        <label htmlFor='password'><i className="fas fa-lock"></i></label>
                        <input name='password' value={this.state.password} id='password' onChange={this.handleChange} placeholder='Password (between 5-10 characters)'/>
                    </div>    
                        { (errorStatus !== '' && errorStatus !== 'Input not successful') && <p>Email has already been used</p>}
                        { (errorStatus === 'Input not successful') && <p>All fields must be filled in to create an account</p>}
                    <button className='create-account-btn'>Create Account</button>
                </form>
            )
        }
    }
}

export const mapStateToProps = (state) => ({
    currentUser: state.currentUser,
    errorStatus: state.errorStatus
});

export const mapDispatchToProps = (dispatch) => ({
    updateError: (message) => dispatch(updateError(message)),
    loginUser: (id, name) => dispatch(loginUser(id, name))
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateAccount);