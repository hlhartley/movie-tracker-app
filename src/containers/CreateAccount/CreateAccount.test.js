import React from 'react';
import { CreateAccount, mapStateToProps, mapDispatchToProps} from './CreateAccount';
import { shallow } from 'enzyme';
import { updateError, loginUser } from '../../actions';
import { mockMovies } from '../../__fixtures__/mockData';
import * as Requests from '../../helpers/requests';
import { validator } from '../../helpers/validators';

describe('CreateAccount', () => {
  describe('CreateAccount component', () => {
    let wrapper;
    const mockLoginUser = jest.fn();
    const mockUpdateError = jest.fn();
    const mockReturnedUser = { id: 5 }
    
    beforeEach(() => {
      let props = {
        currentUser: null,
        errorStatus: ''
      }
      
      wrapper = shallow(<CreateAccount {...props} loginUser={mockLoginUser} updateError={mockUpdateError}/>)
      Requests.createNewUser = jest.fn().mockImplementation(() => Promise.resolve(mockReturnedUser));
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    });

    it('should have a default state', () => {
      const expected = {
        name: '',
        email: '',
        password: '',
      }
      expect(wrapper.state()).toEqual(expected);
    });

    it('should handleChange from the name input', () => {
      const mockEvent = {
        target: {
          name: 'name',
          value: 'Lee'
        }
      }
      const expected = 'Lee'
      wrapper.find('#name').simulate('change', mockEvent)
      expect(wrapper.state('name')).toEqual(expected)
    });

    it('should handleChange from the email input', () => {
      const mockEvent = {
        target: {
          name: 'email',
          value: 'lee@turing.com'
        }
      }
      const expected = 'lee@turing.com'
      wrapper.find('#email').simulate('change', mockEvent)
      expect(wrapper.state('email')).toEqual(expected)
    });

    it('should handleChange from the password input', () => {
      const mockEvent = {
        target: {
          name: 'password',
          value: 'passW@rd'
        }
      }
      const expected = 'passW@rd'
      wrapper.find('#password').simulate('change', mockEvent)
      expect(wrapper.state('password')).toEqual(expected)
    });

    it('should call createNewUser with the correct params', async () => {
      const mockUser = {
        name: 'Sarah',
        email: 'sar3@gmail.com',
        password: 's@rat',
      }
      wrapper.setState(mockUser);
      await wrapper.instance().handleSubmit();
      expect(Requests.createNewUser).toHaveBeenCalledWith(mockUser);

    });

    it('should handleSubmit and login user if everything is okay with createNewUser', async () => {
      const newUserName = 'Sarah'
      wrapper.setState({
        name: newUserName,
        email: 'sar3@gmail.com',
        password: 's@rat',
      });
      await wrapper.instance().handleSubmit();
      expect(mockLoginUser).toHaveBeenCalledWith(5, 'Sarah');
    });

    it('should call updateError with an error message if everything is not okay with createNewUser', async () => {
      Requests.createNewUser = jest.fn(() => {
        throw Error('Email has already been used')
      });

      await wrapper.instance().handleSubmit();
      expect(mockUpdateError).toHaveBeenCalledWith('Email has already been used');
    });

    it('should validateInput and if there are no issues with name email or password will call handleSubmit', () => {
      const mockEvent = {
        preventDefault: jest.fn()
      }

      wrapper.setState({
        name: 'Joe',
        email: 'joe5@gmail.com',
        password: 'joespaSS',
      });
      const instance = wrapper.instance();
      jest.spyOn(instance, 'handleSubmit');
      wrapper.instance().validateInput(mockEvent);
      expect(instance.handleSubmit).toHaveBeenCalled();
    });

    it('should validateInput and if anything is wrong with name email or password will call updateError', () => {
      const mockEvent = {
        preventDefault: jest.fn()
      }
      const expected = 'Input not successful';

      wrapper.setState({
        name: 'Joe',
        email: 'joe5@',
        password: 'joespaSS',
      });

      wrapper.instance().validateInput(mockEvent);
      expect(mockUpdateError).toHaveBeenCalledWith(expected);
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object with an error status', () => {
      const mockState = {
        movie: mockMovies,
        currentUser: {
          id: 2,
          name: 'Sarah'
        },
        errorStatus: ''
      }

      const expected = {
        currentUser: {
          id: 2,
          name: 'Sarah'
        },
        errorStatus: ''
      }

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with a loginUser action when loginUser is called', () => {
      const mockDispatch = jest.fn();
      const id = 3;
      const name = 'Sarah'
      const actionToDispatch = loginUser(id, name);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.loginUser(id, name);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with an updateError action when updateError is called', () => {
      const mockDispatch = jest.fn();
      const message = 'Error creating account';
      const actionToDispatch = updateError(message);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.updateError(message);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })
  });
});