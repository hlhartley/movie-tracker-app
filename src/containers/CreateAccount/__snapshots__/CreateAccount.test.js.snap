// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccount CreateAccount component should match the snapshot 1`] = `
<form
  className="form-container"
  onSubmit={[Function]}
>
  <p
    className="main-text"
  >
    SIGN UP FOR FREE
  </p>
  <p
    className="sub-text"
  >
    Start tracking your favorite movies today!
  </p>
  <div
    className="name-container"
  >
    <label
      htmlFor="name"
    >
      <i
        class="fas fa-user"
      />
    </label>
    <input
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="Full name"
      value=""
    />
  </div>
  <div
    className="email-container"
  >
    <label
      htmlFor="email"
    >
      <i
        class="far fa-envelope"
      />
    </label>
    <input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="E-mail address"
      value=""
    />
  </div>
  <div
    className="password-container"
  >
    <label
      htmlFor="password"
    >
      <i
        class="fas fa-lock"
      />
    </label>
    <input
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="Password (between 5-10 characters)"
      value=""
    />
  </div>
  <button
    className="create-account-btn"
  >
    Create Account
  </button>
</form>
`;
